project-repo/code-practices.md

```diff
@@ -0,0 +1,41 @@
+# Code practices
+
+- All new code comes with tests. Fixes not coming with tests is rather the exception than the norm.
+
+- We follow upstream/community formatting conventions for the given technology. For instance. in Go, we tighten it using `gofumt` instead of `gofmt`. In Rust, `rustfmt`.
+
+- The code should always be linting-error free.
+
+- We should not have any warning/error when building a project. Any of those needs to be "clear" (with a comment explaining why we are ignoring this warning).
+
+- Every method have a docstring/annotation explaining what the function does, what it accepts and what it returns. Also, for languages raising exceptions, if any kind of specific exception that this method raises.
+
+- Two code copies are OK, if you start having a third one, it's probably time to generalize/factor it out.
+
+- Always have private interfaces/modules first, and only open them up if you use them. Do not pre-determine that this element should be public because you will use it one day.
+
+- Avoid using global variables in modules. Prefer a struct that embeds the state and that can be mock for other package level tests.
+
+- Commit convention: something I have a less strict opinion about. I think there should always be a description and a short commit title. The commit title can reference the domain of the project this work is acting on. There is some proposal to use Conventional Commits, how do we relate it to non squash commits if we don't follow this?
```

Comment:
ok, this is definitively a bare copy of my dumped notes on the google doc, shouldn’t we rephrase it first? Also, this evolved (see the PR session our squads gave at Riga + the pull request doc).